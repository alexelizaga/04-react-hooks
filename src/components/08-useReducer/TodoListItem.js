import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ index, todo, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && "complete"}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-danger"
        onClick={() => handleDelete(todo.id)}
      >
        Borrar
      </button>
    </li>
  )
};

TodoListItem.propTypes = {
  index: PropTypes.number.isRequired,
  todo: PropTypes.object.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}
