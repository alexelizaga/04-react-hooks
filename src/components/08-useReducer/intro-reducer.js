
const initialState = [{
  id: 1,
  todo: 'Comprar pan',
  done: false,
}];

const todoReducer = ( state = initialState, action ) => {
  switch (action?.type) {
    case 'add':
      return [ ...state, action.payload ]
  
    default:
      return state;
  }
}
let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'Comprar leche',
  done: false,
}

const action = { type: 'add', payload: newTodo }

todos = todoReducer( todos, action );






console.log(todos);