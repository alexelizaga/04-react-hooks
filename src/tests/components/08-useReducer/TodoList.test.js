import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test <TodoList />', () => {

  const props = {
    todos: demoTodos,
    handleToggle: jest.fn(),
    handleDelete: jest.fn(),
  }

  test('should render correctly', () => {
    const tree = renderer.create(<TodoList {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should have two <TodoListItem />', () => {
    render(<TodoList {...props} />);

    const todoListItem = screen.getAllByTestId('TodoListItem');
    expect(todoListItem.length).toBe(demoTodos.length);
  })

});