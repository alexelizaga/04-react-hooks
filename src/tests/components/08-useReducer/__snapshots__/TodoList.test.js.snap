// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoList /> should render correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item"
    data-testid="TodoListItem"
  >
    <p
      className="false"
      data-testid="TodoListItemDesc"
      onClick={[Function]}
    >
      1
      . 
      Aprender React
    </p>
    <button
      className="btn btn-danger"
      data-testid="TodoListItemBtn"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
  <li
    className="list-group-item"
    data-testid="TodoListItem"
  >
    <p
      className="false"
      data-testid="TodoListItemDesc"
      onClick={[Function]}
    >
      2
      . 
      Aprender Mongo
    </p>
    <button
      className="btn btn-danger"
      data-testid="TodoListItemBtn"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</ul>
`;
